File: library/src/main/java/com/xwray/groupie/Item.java
package com.xwray.groupie;

import androidx.annotation.CallSuper;
import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicLong;

public abstract class Item<VH extends GroupieViewHolder> implements Group, SpanSizeProvider {

    private static AtomicLong ID_COUNTER = new AtomicLong(0);
    protected GroupDataObserver parentDataObserver;
    private final long id;
    private Map<String, Object> extras = new HashMap<>();

    public Item() {
        this(ID_COUNTER.decrementAndGet());
    }

    protected Item(long id) {
        this.id = id;
    }

    @NonNull
    public VH createViewHolder(@NonNull View itemView) {
        return (VH) new GroupieViewHolder(itemView);
    }

    /**
     * Perform any actions required to set up the view for display.
     *
     * @param viewHolder          The viewHolder to bind
     * @param position            The adapter position
     * @param payloads            Any payloads (this list may be empty)
     * @param onItemClickListener An optional adapter-level click listener
     * @param onItemLongClickListener An optional adapter-level long click listener
     */
    @CallSuper
    public void bind(@NonNull VH viewHolder, int position, @NonNull List<Object> payloads,
                     @Nullable OnItemClickListener onItemClickListener,
                     @Nullable OnItemLongClickListener onItemLongClickListener) {
        viewHolder.bind(this, onItemClickListener, onItemLongClickListener);
        bind(viewHolder, position, payloads);
    }

    public abstract void bind(@NonNull VH viewHolder, int position);

    /**
     * If you don't specify how to handle payloads in your implementation, they'll be ignored and
     * the adapter will do a full rebind.
     *
     * @param viewHolder The ViewHolder to bind
     * @param position The adapter position
     * @param payloads A list of payloads (may be empty)
     */
    public void bind(@NonNull VH viewHolder, int position, @NonNull List<Object> payloads) {
        bind(viewHolder, position);
    }

    /**
     * Do any cleanup required for the viewholder to be reused.
     *
     * @param viewHolder The ViewHolder being recycled
     */
    @CallSuper
    public void unbind(@NonNull VH viewHolder) {
        viewHolder.unbind();
    }

    /**
     * Whether the view should be recycled. Return false to prevent the view from being recycled.
     * (Note that it may still be re-bound.)
     *
     * @return Whether the view should be recycled.
     * @see RecyclerView.Adapter#onFailedToRecycleView(RecyclerView.ViewHolder)
     */
    public boolean isRecyclable() {
        return true;
    }

    @Override
    public int getSpanSize(int spanCount, int position) {
        return spanCount;
    }

    public int getSwipeDirs() {
        return 0;
    }

    public int getDragDirs() {
        return 0;
    }

    @LayoutRes
    public abstract int getLayout();

    public void onViewAttachedToWindow(@NonNull VH viewHolder) {}

    public void onViewDetachedFromWindow(@NonNull VH viewHolder) {}

    /**
     * Override this method if the same layout needs to have different viewTypes.
     * @return the viewType, defaults to the layoutId
     * @see RecyclerView.Adapter#getItemViewType(int)
     */
    public int getViewType() {
        return getLayout();
    }

    @Override
    public int getItemCount() {
        return 1;
    }

    @Override
    @NonNull
    public Item getItem(int position) {
        if (position == 0) {
            return this;
        } else {
            throw new IndexOutOfBoundsException("Wanted item at position " + position + " but" +
                    " an Item is a Group of size 1");
        }
    }

    @Override
    public void registerGroupDataObserver(@NonNull GroupDataObserver groupDataObserver) {
        this.parentDataObserver = groupDataObserver;
    }

    @Override
    public void unregisterGroupDataObserver(@NonNull GroupDataObserver groupDataObserver) {
        parentDataObserver = null;
    }

    @Override
    public int getPosition(@NonNull Item item) {
        return this == item ? 0 : -1;
    }

    public boolean isClickable() {
        return true;
    }

    public boolean isLongClickable() {
        return true;
    }

    public void notifyChanged() {
        if (parentDataObserver != null) {
            parentDataObserver.onItemChanged(this, 0);
        }
    }

    public void notifyChanged(@Nullable Object payload) {
        if (parentDataObserver != null) {
            parentDataObserver.onItemChanged(this, 0, payload);
        }
    }

    /**
     * A set of key/value pairs stored on the ViewHolder that can be useful for distinguishing
     * items of the same view type.
     *
     * @return The map of extras
     */
    public Map<String, Object> getExtras() {
        return extras;
    }

    /**
     * If you don't specify an id, this id is an auto-generated unique negative integer for each Item (the less
     * likely to conflict with your model IDs.)
     * <p>
     * You may prefer to override it with the ID of a model object, for example the primary key of
     * an object from a database that it represents.
     *
     * @return A unique id
     */
    public long getId() {
        return id;
    }

    /**
     * Whether two item objects represent the same underlying data when compared using DiffUtil,
     * even if there has been a change in that data.
     * <p>
     * The default implementation compares both view type and id.
     *
     * @return True if the items are the same, false otherwise.
     */
    public boolean isSameAs(@NonNull Item other) {
        if (getViewType() != other.getViewType()) {
            return false;
        }
        return getId() == other.getId();
    }

    /**
     * Whether this item has the same content as another when compared using DiffUtil.
     *
     * After two items have been determined to be the same using {@link #isSameAs(Item)} this function
     * should check whether their contents are the same.
     *
     * The default implementation does this using {@link #equals(Object)}
     *
     * @return True if both items have the same content, false otherwise
     */
    public boolean hasSameContentAs(@NonNull Item other) {
        return this.equals(other);
    }

    @Nullable
    public Object getChangePayload(@NonNull Item newItem) {
        return null;
    }
}

File: library-kotlin/src/main/java/com/wray/groupiekotlin/Item.kt
package com.wray.groupiekotlin

import android.view.View
import androidx.annotation.CallSuper
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView
import java.util.concurrent.atomic.AtomicLong

/**
 * A typealias for `Item<*>` to make it nicer looking.
 */
typealias AnyItem = Item<*>

abstract class Item<VH : GroupieViewHolder>(
    private val _id: Long = ID_COUNTER.decrementAndGet()
): Group, SpanSizeProvider {

    companion object {
        private val ID_COUNTER = AtomicLong(0)
    }

    /**
     * A set of key/value pairs stored on the ViewHolder that can be useful for distinguishing
     * items of the same view type.
     */
    val extras = mutableMapOf<String, Any>()
    // TODO: Add documentation
    open val swipeDirs: Int = 0
    // TODO: Add documentation
    open val dragDirs: Int = 0
    // TODO: Add documentation
    open val isClickable: Boolean = true
    // TODO: Add documentation
    open val isLongClickable: Boolean = true

    @get:LayoutRes
    abstract val layoutRes: Int

    abstract fun bind(viewHolder: VH, position: Int)


    /**
     * Whether the view should be recycled. Return false to prevent the view from being recycled.
     * (Note that it may still be re-bound.)
     *
     * @return Whether the view should be recycled.
     * @see RecyclerView.Adapter.onFailedToRecycleView
     */
    open val isRecyclable: Boolean = true

    private var parentDataObserver: GroupDataObserver? = null

    @Suppress("UNCHECKED_CAST")
    fun createViewHolder(itemView: View): VH = GroupieViewHolder(itemView) as VH

    /**
     * Perform any actions required to set up the view for display.
     *
     * @param viewHolder          The viewHolder to bind
     * @param position            The adapter position
     * @param payloads            Any payloads (this list may be empty)
     * @param onItemClickListener An optional adapter-level click listener
     * @param onItemLongClickListener An optional adapter-level long click listener
     */
    @CallSuper
    open fun bind(viewHolder: VH,
             position: Int,
             payloads: List<Any> = emptyList(),
             onItemClickListener: OnItemClickListener? = null,
             onItemLongClickListener: OnItemLongClickListener? = null) {
        viewHolder.bind(this, onItemClickListener, onItemLongClickListener)
        bind(viewHolder, position, payloads)
    }

    open fun bind(viewHolder: VH, position: Int, payloads: List<Any>) {
        bind(viewHolder, position)
    }

    @CallSuper
    open fun unbind(viewHolder: VH) {
        viewHolder.unbind()
    }

    /**
     * Override this method if the same layout needs to have different viewTypes.
     * @return the viewType, defaults to the layoutId
     * @see RecyclerView.Adapter.getItemViewType(Int)
     */
    val viewType: Int
        get() = layoutRes

    fun notifyChanged() {
        parentDataObserver?.onItemChanged(this, 0)
    }

    fun notifyChanged(payload: Any) {
        parentDataObserver?.onItemChanged(this, 0, payload)
    }

    /**
     * If you don't specify an id, this id is an auto-generated unique negative integer for each Item (the less
     * likely to conflict with your model IDs.)
     *
     *
     * You may prefer to override it with the ID of a model object, for example the primary key of
     * an object from a database that it represents.
     *
     * @return A unique id
     */
    open val id: Long
        get() = _id

    open fun isSameAs(other: AnyItem): Boolean =
        if (viewType != other.viewType) {
            false
        } else id == other.id

    open fun getChangePayload(newItem: Item<VH>): Any? = null

    //region: Group Interface

    override fun registerGroupDataObserver(groupDataObserver: GroupDataObserver) {
        parentDataObserver = groupDataObserver
    }

    override fun unregisterGroupDataObserver(groupDataObserver: GroupDataObserver) {
        parentDataObserver = null
    }

    override fun getItem(position: Int): Item<VH> {
        if (position == 0) {
            return this
        } else {
            throw IndexOutOfBoundsException("Wanted item at position $position but this Item is a Group of size 1")
        }
    }

    override fun getSpanSize(spanCount: Int, position: Int): Int = spanCount

    override fun getPosition(item: AnyItem): Int =
        if (this == item) 0 else -1

    override val itemCount: Int = 1

    //endregion: Group interface
}